src/api/app/models/bs_request.rb

```diff
@@ -705,6 +705,15 @@ def cancelapproval(opts)
     approval_handling(nil, opts)
   end
 
+  def calculate_state_from_reviews
+    return :declined if reviews.any?(&:declined?)
```

Comment:
I would define those as scopes and then call `reviews.declined.exists?` It's more performant since it's all happening in SQL. And it's more speaking.

Reply:
But we load the reviews into memory for the `reverse_each` before. So I kind of doubt it's more performant to go back into the DB for the declined ones. Shall we benchmark? :)

Reply:
Yes, let's benchmark it

Reply:
my pleasure :)

the problem I see is the 2nd select in your solution:

```
  Review Load (0.2ms)  SELECT `reviews`.* FROM `reviews` WHERE `reviews`.`bs_request_id` = 37948
  Review Exists (0.2ms)  SELECT  1 AS one FROM `reviews` WHERE `reviews`.`bs_request_id` = 37948 AND `reviews`.`state` = 'declined' LIMIT 1
```

But let's benchmark it on OBS data:

```
requests=BsRequest.all.to_a; puts Benchmark.measure { requests.each { |r| i=r.reviews.to_a; r.calculate_state_from_reviews_bjoern } }
```

vs
```
requests=BsRequest.all.to_a; puts Benchmark.measure { requests.each { |r| i=r.reviews.to_a; r.calculate_state_from_reviews_coolo } }
```

(obviously this requires a lot of memory in both situations, but we don't care about that - hence I'm loading all requests into memory before the measure to avoid this being a factor.

Now bjoern result: 743.018801  36.429906 779.448707 (946.578434)
Now coolo result: 412.326967  22.553010 434.879977 (510.203396)

An improvement of 46% if prefering ruby over database scope.

I understand that scopes are easier to read - but they are super expensive.

Reply:
This discussion and https://github.com/openSUSE/open-build-service/pull/6389#issuecomment-444378744 make me believe that you guys really need to mistrust the database more. Just because it happens in the database doesn't make it fast.